### PR TITLE
Create core package and refactor data label + datamap

### PR DESCRIPTION
**Pull request by wcmjunior**
## Description of the change

1. Refactor `cyral_datalabel` and `cyral_datamap` data sources and resources to use the abstractions provided in `resource.go`.
2. Move all resources and data sources code to folder `internal`.
3. Create the `core` package to centralize all the code related to the provider itself.
4. Simplifies how to create and register resources with new simplified abstractions.
5. Update golang to `1.20`.

FYI: none of the code smell or security issues pointed by sonarcloud were added by this PR.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [x] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review

- [ ] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] Jira issue referenced in commit message and/or PR title

### Testing

```
terraform-provider-cyral % make
go clean -i github.com/cyralinc/terraform-provider-cyral/...
rm -f terraform-provider-cyral_v4.7.2
rm -rf ./out
rm -rf ~/.terraform.d/plugins/local/terraform/cyral/4.7.2
gofmt -w .
mkdir -p out/
# Build for both MacOS and Linux
GOOS=darwin GOARCH=amd64 go build -o out/darwin_amd64/terraform-provider-cyral_v4.7.2 .
GOOS=darwin GOARCH=arm64 go build -o out/darwin_arm64/terraform-provider-cyral_v4.7.2 .
GOOS=linux GOARCH=amd64 go build -o out/linux_amd64/terraform-provider-cyral_v4.7.2 .
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.7.2/darwin_amd64
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.7.2/darwin_arm64
mkdir -p ~/.terraform.d/plugins/local/terraform/cyral/4.7.2/linux_amd64
cp out/darwin_arm64/terraform-provider-cyral_v4.7.2 ~/.terraform.d/plugins/local/terraform/cyral/4.7.2/darwin_arm64
cp out/linux_amd64/terraform-provider-cyral_v4.7.2 ~/.terraform.d/plugins/local/terraform/cyral/4.7.2/linux_amd64
cp out/darwin_amd64/terraform-provider-cyral_v4.7.2 ~/.terraform.d/plugins/local/terraform/cyral/4.7.2/darwin_amd64
go test github.com/cyralinc/terraform-provider-cyral/... -v -race -timeout 20m
?       github.com/cyralinc/terraform-provider-cyral    [no test files]
?       github.com/cyralinc/terraform-provider-cyral/cyral/core [no test files]
=== RUN   TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue
2023/10/30 11:53:51 [DEBUG] Init client.New
2023/10/30 11:53:51 [DEBUG] TokenSource: &{0xc0000b4f00 {0 0} <nil> 0}
2023/10/30 11:53:51 [DEBUG] End client.New
--- PASS: TestNewClient_WhenTLSSkipVerifyIsEnabled_ThenInsecureSkipVerifyIsTrue (0.00s)
=== RUN   TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse
2023/10/30 11:53:51 [DEBUG] Init client.New
2023/10/30 11:53:51 [DEBUG] TokenSource: &{0xc0000b4f48 {0 0} <nil> 0}
2023/10/30 11:53:51 [DEBUG] End client.New
--- PASS: TestNewClient_WhenTLSSkipVerifyIsDisabled_ThenInsecureSkipVerifyIsFalse (0.00s)
=== RUN   TestNewClient_WhenClientIDIsEmpty_ThenThrowError
2023/10/30 11:53:51 [DEBUG] Init client.New
--- PASS: TestNewClient_WhenClientIDIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenClientSecretIsEmpty_ThenThrowError
2023/10/30 11:53:51 [DEBUG] Init client.New
--- PASS: TestNewClient_WhenClientSecretIsEmpty_ThenThrowError (0.00s)
=== RUN   TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError
2023/10/30 11:53:51 [DEBUG] Init client.New
--- PASS: TestNewClient_WhenControlPlaneIsEmpty_ThenThrowError (0.00s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/client       (cached)
?       github.com/cyralinc/terraform-provider-cyral/cyral/internal/datalabel/classificationrule        [no test files]
?       github.com/cyralinc/terraform-provider-cyral/cyral/internal/integration/confextension   [no test files]
=== RUN   TestAccDatalabelDataSource
=== PAUSE TestAccDatalabelDataSource
=== RUN   TestAccDatalabelResource
=== PAUSE TestAccDatalabelResource
=== CONT  TestAccDatalabelResource
=== CONT  TestAccDatalabelDataSource
--- PASS: TestAccDatalabelResource (13.48s)
--- PASS: TestAccDatalabelDataSource (29.98s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/datalabel   30.656s
=== RUN   TestAccSidecarCftTemplateDataSource
=== PAUSE TestAccSidecarCftTemplateDataSource
=== RUN   TestAccSidecarInstanceIDsDataSource
=== PAUSE TestAccSidecarInstanceIDsDataSource
=== RUN   TestIntegrationsData_GetValue_Default
--- PASS: TestIntegrationsData_GetValue_Default (0.00s)
=== RUN   TestIntegrationsData_GetValue_Splunk
--- PASS: TestIntegrationsData_GetValue_Splunk (0.00s)
=== RUN   TestAccDatadogIntegrationResource
=== PAUSE TestAccDatadogIntegrationResource
=== RUN   TestAccELKIntegrationResource
=== PAUSE TestAccELKIntegrationResource
=== RUN   TestAccIdPIntegrationResource
=== PAUSE TestAccIdPIntegrationResource
=== RUN   TestAccLogstashIntegrationResource
=== PAUSE TestAccLogstashIntegrationResource
=== RUN   TestAccLookerIntegrationResource
=== PAUSE TestAccLookerIntegrationResource
=== RUN   TestAccSplunkIntegrationResource
=== PAUSE TestAccSplunkIntegrationResource
=== RUN   TestAccSumoLogicIntegrationResource
=== PAUSE TestAccSumoLogicIntegrationResource
=== CONT  TestAccELKIntegrationResource
=== CONT  TestAccLookerIntegrationResource
=== CONT  TestAccSidecarInstanceIDsDataSource
=== CONT  TestAccSidecarCftTemplateDataSource
=== CONT  TestAccSplunkIntegrationResource
=== CONT  TestAccSumoLogicIntegrationResource
=== CONT  TestAccLogstashIntegrationResource
=== CONT  TestAccDatadogIntegrationResource
=== CONT  TestAccIdPIntegrationResource
--- PASS: TestAccSumoLogicIntegrationResource (12.36s)
--- PASS: TestAccSplunkIntegrationResource (12.40s)
--- PASS: TestAccLookerIntegrationResource (12.95s)
--- PASS: TestAccELKIntegrationResource (12.98s)
--- PASS: TestAccDatadogIntegrationResource (13.45s)
--- PASS: TestAccSidecarCftTemplateDataSource (18.50s)
--- PASS: TestAccSidecarInstanceIDsDataSource (21.35s)
--- PASS: TestAccLogstashIntegrationResource (21.50s)
--- PASS: TestAccIdPIntegrationResource (68.05s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/deprecated  68.446s
=== RUN   TestIntegrationAWSIAMAuthN
=== PAUSE TestIntegrationAWSIAMAuthN
=== CONT  TestIntegrationAWSIAMAuthN
--- PASS: TestIntegrationAWSIAMAuthN (23.98s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/integration/awsiam  24.873s
=== RUN   TestAccDuoMFAIntegrationResource
=== PAUSE TestAccDuoMFAIntegrationResource
=== CONT  TestAccDuoMFAIntegrationResource
--- PASS: TestAccDuoMFAIntegrationResource (14.45s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/integration/confextension/mfaduo    15.631s
=== RUN   TestAccPagerDutyIntegrationResource
=== PAUSE TestAccPagerDutyIntegrationResource
=== CONT  TestAccPagerDutyIntegrationResource
--- PASS: TestAccPagerDutyIntegrationResource (13.02s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/integration/confextension/pagerduty 14.368s
=== RUN   TestAccHCVaultIntegrationResource
=== PAUSE TestAccHCVaultIntegrationResource
=== CONT  TestAccHCVaultIntegrationResource
--- PASS: TestAccHCVaultIntegrationResource (13.57s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/integration/hcvault 15.168s
=== RUN   TestAccIntegrationIdPSAMLDataSource
=== PAUSE TestAccIntegrationIdPSAMLDataSource
=== RUN   TestAccIntegrationIdPSAMLDraftResource
=== PAUSE TestAccIntegrationIdPSAMLDraftResource
=== RUN   TestAccIntegrationIdPSAMLResource
=== PAUSE TestAccIntegrationIdPSAMLResource
=== CONT  TestAccIntegrationIdPSAMLDraftResource
=== CONT  TestAccIntegrationIdPSAMLResource
=== CONT  TestAccIntegrationIdPSAMLDataSource
--- PASS: TestAccIntegrationIdPSAMLDraftResource (19.42s)
--- PASS: TestAccIntegrationIdPSAMLDataSource (35.51s)
--- PASS: TestAccIntegrationIdPSAMLResource (43.79s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/integration/idpsaml 45.651s
=== RUN   TestAccLoggingIntegrationDataSource
=== PAUSE TestAccLoggingIntegrationDataSource
=== RUN   TestAccLogsIntegrationResourceCloudWatch
=== PAUSE TestAccLogsIntegrationResourceCloudWatch
=== RUN   TestAccLogsIntegrationResourceDataDog
=== PAUSE TestAccLogsIntegrationResourceDataDog
=== RUN   TestAccLogsIntegrationResourceElk
=== PAUSE TestAccLogsIntegrationResourceElk
=== RUN   TestAccLogsIntegrationResourceElkEmptyEsCredentials
=== PAUSE TestAccLogsIntegrationResourceElkEmptyEsCredentials
=== RUN   TestAccLogsIntegrationResourceSplunk
=== PAUSE TestAccLogsIntegrationResourceSplunk
=== RUN   TestAccLogsIntegrationResourceSumologic
=== PAUSE TestAccLogsIntegrationResourceSumologic
=== RUN   TestAccLogsIntegrationResourceFluentbit
=== PAUSE TestAccLogsIntegrationResourceFluentbit
=== CONT  TestAccLoggingIntegrationDataSource
=== CONT  TestAccLogsIntegrationResourceFluentbit
=== CONT  TestAccLogsIntegrationResourceElk
=== CONT  TestAccLogsIntegrationResourceDataDog
=== CONT  TestAccLogsIntegrationResourceCloudWatch
=== CONT  TestAccLogsIntegrationResourceSplunk
=== CONT  TestAccLogsIntegrationResourceElkEmptyEsCredentials
=== CONT  TestAccLogsIntegrationResourceSumologic
--- PASS: TestAccLogsIntegrationResourceSumologic (14.56s)
--- PASS: TestAccLogsIntegrationResourceDataDog (14.68s)
--- PASS: TestAccLogsIntegrationResourceFluentbit (14.69s)
--- PASS: TestAccLogsIntegrationResourceElkEmptyEsCredentials (14.77s)
--- PASS: TestAccLogsIntegrationResourceCloudWatch (16.01s)
--- PASS: TestAccLogsIntegrationResourceSplunk (16.26s)
--- PASS: TestAccLogsIntegrationResourceElk (16.37s)
--- PASS: TestAccLoggingIntegrationDataSource (18.19s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/integration/logging 20.139s
=== RUN   TestAccSlackAlertsIntegrationResource
=== PAUSE TestAccSlackAlertsIntegrationResource
=== CONT  TestAccSlackAlertsIntegrationResource
--- PASS: TestAccSlackAlertsIntegrationResource (14.88s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/integration/slack   16.321s
=== RUN   TestAccMsTeamsIntegrationResource
=== PAUSE TestAccMsTeamsIntegrationResource
=== CONT  TestAccMsTeamsIntegrationResource
--- PASS: TestAccMsTeamsIntegrationResource (14.44s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/integration/teams   16.042s
=== RUN   TestAccPermissionDataSource
=== PAUSE TestAccPermissionDataSource
=== CONT  TestAccPermissionDataSource
--- PASS: TestAccPermissionDataSource (7.19s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/permission  7.523s
=== RUN   TestAccPolicyResource
=== PAUSE TestAccPolicyResource
=== CONT  TestAccPolicyResource
--- PASS: TestAccPolicyResource (7.17s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/policy      7.509s
=== RUN   TestAccPolicyRuleResource
=== PAUSE TestAccPolicyRuleResource
=== RUN   TestPolicyRuleResourceUpgradeV0
--- PASS: TestPolicyRuleResourceUpgradeV0 (0.00s)
=== CONT  TestAccPolicyRuleResource
--- PASS: TestAccPolicyRuleResource (19.24s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/policy/rule 19.577s
=== RUN   TestAccRegoPolicyInstanceResource
=== PAUSE TestAccRegoPolicyInstanceResource
=== CONT  TestAccRegoPolicyInstanceResource
--- PASS: TestAccRegoPolicyInstanceResource (7.99s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/regopolicy  8.306s
=== RUN   TestAccRepositoryDataSource
=== PAUSE TestAccRepositoryDataSource
=== RUN   TestAccRepositoryResource
=== PAUSE TestAccRepositoryResource
=== CONT  TestAccRepositoryResource
=== CONT  TestAccRepositoryDataSource
--- PASS: TestAccRepositoryDataSource (15.98s)
--- PASS: TestAccRepositoryResource (21.65s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/repository  21.958s
=== RUN   TestAccRepositoryAccessGatewayResource
=== PAUSE TestAccRepositoryAccessGatewayResource
=== CONT  TestAccRepositoryAccessGatewayResource
--- PASS: TestAccRepositoryAccessGatewayResource (22.83s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/repository/accessgateway    23.154s
=== RUN   TestAccRepositoryAccessRulesResource
=== PAUSE TestAccRepositoryAccessRulesResource
=== CONT  TestAccRepositoryAccessRulesResource
--- PASS: TestAccRepositoryAccessRulesResource (13.67s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/repository/accessrules      14.002s
=== RUN   TestAccRepositoryBindingResource
=== PAUSE TestAccRepositoryBindingResource
=== CONT  TestAccRepositoryBindingResource
--- PASS: TestAccRepositoryBindingResource (15.40s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/repository/binding  15.729s
=== RUN   TestAccRepositoryConfAnalysisResource
=== PAUSE TestAccRepositoryConfAnalysisResource
=== RUN   TestRepositoryConfAnalysisResourceUpgradeV0
--- PASS: TestRepositoryConfAnalysisResourceUpgradeV0 (0.00s)
=== CONT  TestAccRepositoryConfAnalysisResource
--- PASS: TestAccRepositoryConfAnalysisResource (10.26s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/repository/confanalysis     10.582s
=== RUN   TestAccRepositoryConfAuthResource
=== PAUSE TestAccRepositoryConfAuthResource
=== RUN   TestRepositoryConfAuthResourceUpgradeV0
--- PASS: TestRepositoryConfAuthResourceUpgradeV0 (0.00s)
=== CONT  TestAccRepositoryConfAuthResource
--- PASS: TestAccRepositoryConfAuthResource (14.65s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/repository/confauth 14.985s
=== RUN   TestAccRepositoryDatamapResource
=== PAUSE TestAccRepositoryDatamapResource
=== CONT  TestAccRepositoryDatamapResource
--- PASS: TestAccRepositoryDatamapResource (22.77s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/repository/datamap  23.092s
=== RUN   TestAccRepositoryNetworkAccessPolicyResource
=== PAUSE TestAccRepositoryNetworkAccessPolicyResource
=== CONT  TestAccRepositoryNetworkAccessPolicyResource
--- PASS: TestAccRepositoryNetworkAccessPolicyResource (25.94s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/repository/network  26.260s
=== RUN   TestAccRepositoryUserAccountResource
=== PAUSE TestAccRepositoryUserAccountResource
=== CONT  TestAccRepositoryUserAccountResource
--- PASS: TestAccRepositoryUserAccountResource (42.97s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/repository/useraccount      43.337s
=== RUN   TestAccRoleDataSource
=== PAUSE TestAccRoleDataSource
=== RUN   TestAccRoleSSOGroupsResource
=== PAUSE TestAccRoleSSOGroupsResource
=== RUN   TestRoleSSOGroupsResourceUpgradeV0
--- PASS: TestRoleSSOGroupsResourceUpgradeV0 (0.00s)
=== RUN   TestAccRoleResource
=== PAUSE TestAccRoleResource
=== CONT  TestAccRoleDataSource
=== CONT  TestAccRoleSSOGroupsResource
=== CONT  TestAccRoleResource
--- PASS: TestAccRoleDataSource (21.36s)
--- PASS: TestAccRoleSSOGroupsResource (24.58s)
--- PASS: TestAccRoleResource (30.97s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/role        31.308s
=== RUN   TestAccSAMLCertificateDataSource
=== PAUSE TestAccSAMLCertificateDataSource
=== CONT  TestAccSAMLCertificateDataSource
--- PASS: TestAccSAMLCertificateDataSource (5.18s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/samlcertificate     5.500s
=== RUN   TestAccSAMLConfigurationDataSource
=== PAUSE TestAccSAMLConfigurationDataSource
=== CONT  TestAccSAMLConfigurationDataSource
--- PASS: TestAccSAMLConfigurationDataSource (15.24s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/samlconfiguration   15.568s
=== RUN   TestAccServiceAccountResource
=== PAUSE TestAccServiceAccountResource
=== CONT  TestAccServiceAccountResource
--- PASS: TestAccServiceAccountResource (26.24s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/serviceaccount      26.593s
=== RUN   TestAccSidecarBoundPortsDataSource
=== PAUSE TestAccSidecarBoundPortsDataSource
=== RUN   TestAccSidecarIDDataSource
=== PAUSE TestAccSidecarIDDataSource
=== RUN   TestAccSidecarResource
=== PAUSE TestAccSidecarResource
=== CONT  TestAccSidecarIDDataSource
=== CONT  TestAccSidecarResource
=== CONT  TestAccSidecarBoundPortsDataSource
--- PASS: TestAccSidecarIDDataSource (12.69s)
--- PASS: TestAccSidecarBoundPortsDataSource (19.60s)
--- PASS: TestAccSidecarResource (30.51s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/sidecar     30.840s
=== RUN   TestAccSidecarCredentialsResource
=== PAUSE TestAccSidecarCredentialsResource
=== CONT  TestAccSidecarCredentialsResource
--- PASS: TestAccSidecarCredentialsResource (11.33s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/sidecar/credentials 11.659s
=== RUN   TestAccSidecarHealthDataSource
=== PAUSE TestAccSidecarHealthDataSource
=== CONT  TestAccSidecarHealthDataSource
--- PASS: TestAccSidecarHealthDataSource (10.62s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/sidecar/health      10.933s
=== RUN   TestAccSidecarInstanceStatsDataSource
=== PAUSE TestAccSidecarInstanceStatsDataSource
=== RUN   TestAccSidecarInstanceDataSource
=== PAUSE TestAccSidecarInstanceDataSource
=== CONT  TestAccSidecarInstanceStatsDataSource
=== CONT  TestAccSidecarInstanceDataSource
--- PASS: TestAccSidecarInstanceStatsDataSource (9.51s)
--- PASS: TestAccSidecarInstanceDataSource (11.08s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/sidecar/instance    11.423s
=== RUN   TestAccSidecarListenerDataSource
=== PAUSE TestAccSidecarListenerDataSource
=== RUN   TestSidecarListenerResource
=== PAUSE TestSidecarListenerResource
=== CONT  TestSidecarListenerResource
=== CONT  TestAccSidecarListenerDataSource
--- PASS: TestAccSidecarListenerDataSource (15.81s)
--- PASS: TestSidecarListenerResource (29.61s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/sidecar/listener    29.975s
testing: warning: no tests to run
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/sweep       (cached) [no tests to run]
=== RUN   TestAccSystemInfoDataSource
=== PAUSE TestAccSystemInfoDataSource
=== CONT  TestAccSystemInfoDataSource
--- PASS: TestAccSystemInfoDataSource (4.98s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/internal/systeminfo  5.497s
=== RUN   TestAccProvider
2023/11/02 00:25:41 [DEBUG] Init getDataSourceMap
2023/11/02 00:25:41 [DEBUG] Looking for datasources in package `datalabel`
2023/11/02 00:25:41 [DEBUG] Registering datasources `cyral_datalabel`
2023/11/02 00:25:41 [DEBUG] Looking for datasources in package `datamap`
2023/11/02 00:25:41 [DEBUG] Init dataSourceSidecarListener
2023/11/02 00:25:41 [DEBUG] End dataSourceSidecarListener
2023/11/02 00:25:41 [DEBUG] end getDataSourceMap
2023/11/02 00:25:41 [DEBUG] Init getResourceMap
2023/11/02 00:25:41 [DEBUG] Looking for resources in package `datalabel`
2023/11/02 00:25:41 [DEBUG] Registering resources `cyral_datalabel`
2023/11/02 00:25:41 [DEBUG] Looking for resources in package `datamap`
2023/11/02 00:25:41 [DEBUG] Registering resources `cyral_repository_datamap`
2023/11/02 00:25:41 [DEBUG] End getResourceMap
--- PASS: TestAccProvider (0.00s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/provider     0.259s
=== RUN   TestElementsMatch
--- PASS: TestElementsMatch (0.00s)
PASS
ok      github.com/cyralinc/terraform-provider-cyral/cyral/utils        (cached)```

### CODE REVIEW

File: src/cyral/datalabel/resource_test.go [hunk 1]
```diff
@@ -42,9 +45,12 @@ func TestAccDatalabelResource(t *testing.T) {
 		"main_test", initialDataLabelConfig())
 	testUpdatedConfig, testUpdatedFunc := setupDatalabelTest(t,
 		"main_test", updatedDataLabelConfig())
-
 	resource.ParallelTest(t, resource.TestCase{
-		ProviderFactories: providerFactories,
+		ProviderFactories: map[string]func() (*schema.Provider, error){
```
Comment (VictorGFM) [inline src/cyral/datalabel/resource_test.go:49]: Why not keep using the `providerFactories`? I believe Im missing something

**Review by VictorGFM — APPROVED** [review]
@wcmjunior Great refactoring! I just have some NIT comments below, but Im already approving the PR.

File: src/core/README.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+// datasource.go
+
+func init() {
+	sr := &core.SchemaRegister{
+        // Data source name
+		Name:   "cyral_datalabel",
+        // this is a function that will return the actual schema
+		Schema: ResourceSchema,
+        // Schema type
+		Type:   core.ResourceSchema,
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply src/core/README.md]: Addressed in [b69a7c4](https://github.com/cyralinc/terraform-provider-cyral/pull/439/commits/b69a7c44d1a9f5ad5f25588049a06d35ef8d5479).

File: src/utils/testutils.go [hunk 3]
```diff
@@ -10,68 +10,50 @@ import (
...
 //
 // Example usage for cyral_datalabel resource:
 //
-//	accTestName("datalabel", "label1")
+//	A("datalabel", "label1")
 //
 // Example usage for cyral_datalabel data source:
 //
-//	accTestName("data-datalabel", "label1")
+//	A("data-datalabel", "label1")
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply src/utils/testutils.go]: Addressed in [b69a7c4](https://github.com/cyralinc/terraform-provider-cyral/pull/439/commits/b69a7c44d1a9f5ad5f25588049a06d35ef8d5479).

File: src/utils/testutils.go [hunk 4]
```diff
@@ -10,68 +10,50 @@ import (
...
+// A attempts to make resource names unique to a specific resource
 // type, and avoid name clashes with other resources that exist in the testing
 // control plane.
 //
 // Use this for every resource for which name clashes may occur.
 //
 // Example usage for cyral_datalabel resource:
 //
-//	accTestName("datalabel", "label1")
+//	A("datalabel", "label1")
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply src/utils/testutils.go]: Addressed in [b69a7c4](https://github.com/cyralinc/terraform-provider-cyral/pull/439/commits/b69a7c44d1a9f5ad5f25588049a06d35ef8d5479).

File: src/cyral/datalabel/classificationrule/constants.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+const (
+	Unknown = Type("UNKNOWN")
+	Rego    = Type("REGO")
+)
+
+type Status string
+
+const (
+	Enabled  = "ENABLED"
+	Disabled = "DISABLED"
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply src/cyral/datalabel/classificationrule/constants.go]: Addressed in [b69a7c4](https://github.com/cyralinc/terraform-provider-cyral/pull/439/commits/b69a7c44d1a9f5ad5f25588049a06d35ef8d5479).

File: src/cyral/datalabel/resource.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+				Type:     schema.TypeSet,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"rule_type": {
+							Description: "Type of the classification rule. Valid values are: `UNKNOWN` and `REGO`. Defaults " +
+								"to `UNKNOWN`.",
+							Type:         schema.TypeString,
+							Optional:     true,
+							Default:      "UNKNOWN",
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply src/cyral/datalabel/resource.go]: Addressed in [b69a7c4](https://github.com/cyralinc/terraform-provider-cyral/pull/439/commits/b69a7c44d1a9f5ad5f25588049a06d35ef8d5479).

File: src/cyral/datalabel/resource.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+								"REGO code for `REGO`-type classification rules.",
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"rule_status": {
+							Description: "Status of the classification rule. Valid values are: `ENABLED` and  `DISABLED`. " +
+								"Defaults to `ENABLED`.",
+							Type:         schema.TypeString,
+							Optional:     true,
+							Default:      "ENABLED",
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply src/cyral/datalabel/resource.go]: Addressed in [b69a7c4](https://github.com/cyralinc/terraform-provider-cyral/pull/439/commits/b69a7c44d1a9f5ad5f25588049a06d35ef8d5479).

**Comment by wcmjunior** [comment]
@VictorGFM I moved the PR to draft state again as it passes our tests, but it fail actual execution with the resources  and data sources refactored. It is a complex issue, but I will try to summarize as explained below. In order to understand the issue, assume that the package initialization order in golang is determined automatically and it is based on the dependencies explicitly declared, thus if `packageA` depends on `packageB`, `packageB` will be initialized before `packageA`. 

Per the current code, the resources in the new packages `datalabel` and `repository_datamap` do not refer directly to the `Provider` declaration. This means that during the regular plugin initialization using `main.go`, the package `cyral` will be initialized *before* packages `datalabel` and `repository/datamap`. It means that the initialization function that registers these resources to the provider will not be called. =/

The code works during test execution as the tests are initialized individually, thus the initialization order will change and bring packages `datalabel` and `respository/datamap` before `cyral`.

**Comment by VictorGFM** [comment]
@wcmjunior got it. I believe the problem is that the `datalabel` and `datamap` packages are never imported, thus, their init functions will never be called. If we import them, lets say, by creating an `init.go` file that is responsible for importing every resource/datasource package, then this seems to work! Something like:
```Go
package main

// These packages must be imported here so that their respective init functions are called
// and the corresponding resources and data sources defined in each package are registered
// in the provider.
import (
	_ "github.com/cyralinc/terraform-provider-cyral/src/cyral/datalabel"
	_ "github.com/cyralinc/terraform-provider-cyral/src/cyral/repository/datamap"
)
```

**Comment by wcmjunior** [comment]
Yeah, I considered this an option, but I'm trying to find a more elegant solution. The `terraform-provider-aws` uses a curious factory method structure that I would like to investigate more and see if it would fix this problem without becoming too complex. I will get back to this one again in my free time. :)

**Comment by wcmjunior** [comment]
@VictorGFM , final version of the source code is out. I just need to review the `README.md` and I'm done. Feel free to review the code.

File: src/cyral/repository/datamap/resource_test.go [hunk 8]
```diff
@@ -104,7 +108,11 @@ func TestAccRepositoryDatamapResource(t *testing.T) {
 	importStateResName := "cyral_repository_datamap.test_with_datalabel"
 
 	resource.ParallelTest(t, resource.TestCase{
-		ProviderFactories: providerFactories,
+		ProviderFactories: map[string]func() (*schema.Provider, error){
```
[the remark this answers is not in the document]
Comment (wcmjunior) [reply src/cyral/repository/datamap/resource_test.go]: `providerFactories` variable is back.

**Review by VictorGFM — APPROVED** [review]
The overall changes look good to me, thanks for this refactoring.

**Comment by VictorGFM** [comment]
I looked into the SonarCloud warnings but they are just complaining about dummy IPs used for tests, would be good to find a way to ignore these type of warnings.

**Pull request merged**